src/Orleans/EventSourcing/JournaledGrainState.cs

```diff
@@ -28,36 +28,58 @@ MIT License
 
 namespace Orleans.EventSourcing
 {
-    public abstract class JournaledGrainState : GrainState
+    public abstract class JournaledGrainState<TGrainState> : GrainState
+        where TGrainState : JournaledGrainState<TGrainState>
     {
         private List<StateEvent> events = new List<StateEvent>();
-        protected JournaledGrainState(Type type )
-            : base(type.FullName)
+
+        protected JournaledGrainState()
+            : base(typeof(TGrainState).FullName)
         {
         }
 
         public IEnumerable<StateEvent> Events
         {
             get { return events.AsReadOnly(); }
-            set { events = new List<StateEvent>(value); }
         }
 
         public int Version { get; private set; }
 
-        public void AddEvent(StateEvent @event)
+        public void AddEvent<TEvent>(TEvent @event)
+            where TEvent: StateEvent
         {
             events.Add(@event);
-            ApplyEvent(@event);
+
+            DynamicStateTransition(@event);
+            //StaticStateTransition(@event);
+
             Version++;
         }
 
-        public abstract void ApplyEvent(StateEvent @event);
-
         public override void SetAll(IDictionary<string, object> values)
         {
             base.SetAll(values);
-            foreach(StateEvent @event in Events)
-                ApplyEvent(@event);
+
+            dynamic me = this;
+
+            foreach (StateEvent @event in Events)
+                me.Apply(@event);
+        }
+
+        private void DynamicStateTransition<TEvent>(TEvent @event)
+            where TEvent : StateEvent
+        {
+            dynamic me = this;
+            me.Apply(this, @event);
```

Comment:
Not used dynamic enough to speak from experience, but my understanding is that dynamic per event would not be cheep (as compared to a switch, or other methods).

If a new TEvent type was added without a corresponding handler, the code would add the bad event to the event list, but fail to adjust the state, as well as causing the AddEvent to throw.  If the events were eventually persisted somewhere, this could cause some icky bugs.

Suggest considering some sort of handler registration?  Shouldn't add much to developer overhead, while allowing for better error handling as well as faster event handling.